pkg/kv/kvserver/logstore/logstore.go

```diff
@@ -239,7 +239,15 @@ func (s *LogStore) storeEntriesAndCommitBatch(
 	// (Replica), so this comment might need to move.
 	stats.PebbleBegin = crtime.NowMono()
 	stats.PebbleBytes = int64(batch.Len())
-	wantsSync := m.MustSync()
+	// We want a timely sync in two cases:
+	//	1. There are raft messages (e.g. MsgVoteResp and MsgAppResp) conditional
+	//	on this write being durable. Sending these messages is on the critical path
+	//	for writes/consensus.
+	//	2. The log append overwrites some entries. Since there can be sideloaded
+	//	entries to be removed, we must sync before doing so. This is an internal
+	//	detail here: we could instead do a non-blocking sync and remove sideloaded
+	//	entries asynchronously. See #136416.
+	wantsSync := m.MustSync() || overwriting
```

Comment:
Ideally, need a test that makes these sync decisions visible. Will be possible only after the `logstore` refactor, which I'm doing in background.